# Project 1

Project 1: Adolescent fertility rate (births per 1,000 women ages 15-19) Data from the World Bank

In [1]:
import plotly.io as pio

pio.renderers.default = "vscode+jupyterlab+notebook_connected"

In [1]:
#Read in the data
import pandas as pd
df = pd.read_csv('/Users/josefinadesolminihac/Desktop/computing/GenderIndicators.csv', skiprows=4)
df.head()

Country Name Country Code  \
0        Aruba          ABW   
1        Aruba          ABW   
2        Aruba          ABW   
3        Aruba          ABW   
4        Aruba          ABW   

                                      Indicator Name     Indicator Code  1960  \
0  Teenage mothers (% of women ages 15-19 who hav...     SP.MTR.1519.ZS   NaN   
1  Women who were first married by age 18 (% of w...  SP.M18.2024.FE.ZS   NaN   
2  Women who were first married by age 15 (% of w...  SP.M15.2024.FE.ZS   NaN   
3  Female headed households (% of households with...     SP.HOU.FEMA.ZS   NaN   
4           Wanted fertility rate (births per woman)        SP.DYN.WFRT   NaN   

   1961  1962  1963  1964  1965  ...  2015  2016  2017  2018  2019  2020  \
0   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
2   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
3   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
4   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   

   2021  2022  2023  Unnamed: 68  
0   NaN   NaN   NaN          NaN  
1   NaN   NaN   NaN          NaN  
2   NaN   NaN   NaN          NaN  
3   NaN   NaN   NaN          NaN  
4   NaN   NaN   NaN          NaN  

[5 rows x 69 columns]

In [3]:
#Identify the names of the columns so it is easier to analyze 
df.columns

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022',
       '2023', 'Unnamed: 68'],
      dtype='object')

In [5]:
# Filter for the indicator Adolescent fertility rate (births per 1,000 women ages 15-19)
indicator_name = "Adolescent fertility rate (births per 1,000 women ages 15-19)"
filtered_df = df[df['Indicator Name'] == indicator_name]
filtered_df

Country Name Country Code  \
13                           Aruba          ABW   
169    Africa Eastern and Southern          AFE   
325                    Afghanistan          AFG   
481     Africa Western and Central          AFW   
637                         Angola          AGO   
...                            ...          ...   
40729                       Kosovo          XKX   
40885                  Yemen, Rep.          YEM   
41041                 South Africa          ZAF   
41197                       Zambia          ZMB   
41353                     Zimbabwe          ZWE   

                                          Indicator Name Indicator Code  \
13     Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   
169    Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   
325    Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   
481    Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   
637    Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   
...                                                  ...            ...   
40729  Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   
40885  Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   
41041  Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   
41197  Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   
41353  Adolescent fertility rate (births per 1,000 wo...    SP.ADO.TFRT   

             1960        1961        1962        1963        1964        1965  \
13      62.182000   61.589000   61.170000   59.658000   58.324000   56.291000   
169    140.180526  140.810248  141.664168  142.324951  143.456933  144.331150   
325    138.876000  138.717000  138.494000  138.173000  140.107000  140.935000   
481    167.054250  170.748419  174.212286  175.665650  175.847602  175.467021   
637    148.279000  150.275000  150.801000  153.002000  155.661000  156.336000   
...           ...         ...         ...         ...         ...         ...   
40729   81.772000   82.849000   82.555000   81.527000   80.337000   79.408000   
40885  142.832000  144.901000  146.197000  145.273000  145.403000  145.899000   
41041  101.605000  101.186000  101.666000  101.630000  101.854000  101.095000   
41197  184.021000  185.311000  186.542000  187.872000  188.592000  188.400000   
41353  168.191000  167.318000  165.582000  163.887000  157.860000  151.588000   

       ...        2015        2016        2017        2018        2019  \
13     ...   28.989000   27.419000   24.888000   20.086000   18.906000   
169    ...  105.501833  103.222825  100.963526   99.491703   99.085572   
325    ...   97.415000   94.018000   90.342000   88.431000   86.803000   
481    ...  117.728385  113.411758  110.863328  109.930971  108.778536   
637    ...  148.074000  146.135000  145.235000  142.575000  141.278000   
...    ...         ...         ...         ...         ...         ...   
40729  ...   10.122000    9.221000    9.341000    8.847000    8.113000   
40885  ...   64.753000   62.113000   60.335000   58.900000   57.830000   
41041  ...   78.367000   70.813000   63.706000   59.361000   64.982000   
41197  ...  128.861000  129.067000  128.315000  124.849000  122.836000   
41353  ...  102.084000  100.884000   98.265000   98.507000   97.354000   

             2020        2021        2022  2023  Unnamed: 68  
13      16.522000   14.154000   13.697000   NaN          NaN  
169     97.413585   96.181442   94.688181   NaN          NaN  
325     84.296000   82.565000   79.735000   NaN          NaN  
481    107.580169  106.548068  104.671712   NaN          NaN  
637    139.825000  138.402000  135.792000   NaN          NaN  
...           ...         ...         ...   ...          ...  
40729    8.187000    8.225000    8.047000   NaN          NaN  
40885   56.160000   54.404000   52.497000   NaN          NaN  
41041   62.742000   61.206000   60.891000   NaN          NaN  
41197  119.00900

In [7]:
# Filter for the year 2022
filtered_df_2022 = filtered_df[['Country Name','Country Code','2022']].reset_index(drop=True)
filtered_df_2022.head()

Country Name Country Code        2022
0                        Aruba          ABW   13.697000
1  Africa Eastern and Southern          AFE   94.688181
2                  Afghanistan          AFG   79.735000
3   Africa Western and Central          AFW  104.671712
4                       Angola          AGO  135.792000

In [9]:
#See the type of the number column in this case column 2022
filtered_df_2022.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 266 entries, 0 to 265
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Country Name  266 non-null    object 
 1   Country Code  266 non-null    object 
 2   2022          265 non-null    float64
dtypes: float64(1), object(2)
memory usage: 6.4+ KB


In [11]:
#See what we have in 2022 column
filtered_df_2022['2022'].unique()

array([ 13.697     ,  94.68818081,  79.735     , 104.6717119 ,
       135.792     ,  14.46      ,   5.91      ,  44.62252854,
         2.844     ,  37.932     ,  18.305     ,  29.989     ,
        32.262     ,   7.731     ,   5.243     ,  40.081     ,
        52.595     ,   5.105     ,  90.835     , 108.692     ,
        73.345     ,  38.215     ,   8.714     ,  25.079     ,
         9.384     ,  11.191     ,  56.569     ,   7.429     ,
        63.073     ,  43.582     ,  41.864     ,   9.524     ,
        18.481     ,  48.343     , 158.998     ,   6.618     ,
        18.43403203,   2.2       ,  10.737     ,  22.793     ,
        11.146     , 103.293     , 108.618     , 107.481     ,
       101.197     ,  57.612     ,  56.081     ,  53.987     ,
        35.749     ,  50.1176937 ,  48.914     ,  22.722     ,
        10.893     ,   6.825     ,   9.364     ,   7.186     ,
        22.675     ,  37.933     ,   1.814     ,  63.198     ,
        11.564     ,  21.49783636,  34.94036006,  20.10

In [13]:
#Drop all nan values
filtered_df_2022 = filtered_df_2022.dropna()
filtered_df_2022.isna().sum()

Country Name    0
Country Code    0
2022            0
dtype: int64

In [15]:
#Compute: the mean, median, and mode
mean2022 = filtered_df_2022['2022'].mean()
median2022 = filtered_df_2022['2022'].median()
mode2022 = filtered_df_2022['2022'].mode()
mode2022

0    27.851150
1    98.723416
Name: 2022, dtype: float64

In [17]:
import csv

data = []
indicator_name = "Adolescent fertility rate (births per 1,000 women ages 15-19)"
with open('/Users/josefinadesolminihac/Desktop/computing/GenderIndicators.csv', mode='r', encoding='utf-8-sig') as file:
    reader = csv.reader(file)
    
    # Skip the first 4 rows because they have the title or are empty
    for _ in range(4):
        next(reader)
    
    reader = csv.DictReader(file)
    
    # Filter the data based on the indicator name and collect 2022 values
    for row in reader:
        if row.get('Indicator Name') == indicator_name:
            value_str = row['2022']
            if value_str != '':   #had a problem when changing the type to float with the " " value, so I need to skip it
                value = float(value_str)
                data.append(value)

data

[13.697,
 94.6881808086008,
 79.735,
 104.671711904197,
 135.792,
 14.46,
 5.91,
 44.6225285359955,
 2.844,
 37.932,
 18.305,
 29.989,
 32.262,
 7.731,
 5.243,
 40.081,
 52.595,
 5.105,
 90.835,
 108.692,
 73.345,
 38.215,
 8.714,
 25.079,
 9.384,
 11.191,
 56.569,
 7.429,
 63.073,
 43.582,
 41.864,
 9.524,
 18.481,
 48.343,
 158.998,
 6.618,
 18.434032025962,
 2.2,
 10.737,
 22.793,
 11.146,
 103.293,
 108.618,
 107.481,
 101.197,
 57.612,
 56.081,
 53.987,
 35.749,
 50.1176937037383,
 48.914,
 22.722,
 10.893,
 6.825,
 9.364,
 7.186,
 22.675,
 37.933,
 1.814,
 63.198,
 11.564,
 21.497836357712,
 34.9403600574987,
 20.1018360800387,
 19.4406930023359,
 13.1305941154449,
 62.065,
 43.603,
 6.98503849403079,
 63.639,
 6.349,
 8.1,
 66.502,
 9.06646346314163,
 88.396878339882,
 4.1,
 26.141,
 9.442,
 9.697,
 35.692,
 89.807,
 9.985,
 29.693,
 63.41,
 11.911,
 112.231,
 60.681,
 85.775,
 136.439,
 8.315,
 32.141,
 35.017,
 63.163,
 33.869,
 64.636,
 11.1975434180883,
 1.571,
 71.333,
 99.

In [19]:
from statistics import mean, median, multimode
mean_data = mean(data)
print(mean_data)

42.32390601657742


The mean adolescent fertility rate is approximately 42.32 births per 1,000 women aged 15-19. This means that, on average, there are about 42 births per 1,000 adolescent women in this age group.

In [22]:
median_data = median(data)
print(median_data)

33.265


The median adolescent fertility rate is 33.27 births per 1,000 women aged 15-19. This suggests that half of the values in the dataset are below 33.27, and the other half are above, showing the central point of the data distribution.

In [25]:
mode_data = multimode(data)
print(mode_data)

[27.8511499620514, 98.7234155145699]


The modes of the adolescent fertility rate are 27.85 and 98.72 births per 1,000 women aged 15-19. These values represent the most frequently occurring fertility rates in the dataset, indicating that these rates are observed more often than other values.

In [28]:
#For the visualization, I want to group the countries by quartile. 
import numpy as np

data = filtered_df_2022['2022']
#Find the quartiles
q1 = np.quantile(data, 0.25)
q2 = np.quantile(data, 0.5)
q3 = np.quantile(data, 0.75)
print(q1,q2,q3)

11.911 33.265 62.61


In [30]:
#Create a function that assigns the quartile
def assign_quartile(row):
    if row["2022"] <= q1:
        return "Q1"
    elif q1 < row["2022"] <= q2:
        return "Q2"
    elif q2 < row["2022"] <= q3:
        return "Q3"
    else:
        return "Q4"

In [32]:
# Apply the function to the DataFrame
filtered_df_2022["Quartile"] = filtered_df_2022.apply(assign_quartile, axis=1)
filtered_df_2022

Country Name Country Code        2022 Quartile
0                          Aruba          ABW   13.697000       Q2
1    Africa Eastern and Southern          AFE   94.688181       Q4
2                    Afghanistan          AFG   79.735000       Q4
3     Africa Western and Central          AFW  104.671712       Q4
4                         Angola          AGO  135.792000       Q4
..                           ...          ...         ...      ...
261                       Kosovo          XKX    8.047000       Q1
262                  Yemen, Rep.          YEM   52.497000       Q3
263                 South Africa          ZAF   60.891000       Q3
264                       Zambia          ZMB  116.091000       Q4
265                     Zimbabwe          ZWE   92.566000       Q4

[265 rows x 4 columns]

In [41]:
quartile_data = filtered_df_2022.groupby(['Quartile'], as_index=False)['2022'].mean().round(2)
quartile_data

Quartile   2022
0       Q1   7.05
1       Q2  22.80
2       Q3  45.83
3       Q4  94.14

In [36]:
# Calculate the max value in the "2022" column to use as the reference for scaling
max_value = quartile_data["2022"].max()

# Define a function to create the sparkline scale with '*', using a scale of 10 (10 '*' is the biggest number)
def create_sparkline(value, max_value, scale_factor=10):
    num_stars = int((value / max_value) * scale_factor)
    return "*" * num_stars

# Apply the function to each value in the "2022" column
quartile_data["sparkline"] = quartile_data["2022"].apply(create_sparkline, max_value=max_value)

# Display the resulting DataFrame
quartile_data

Quartile   2022   sparkline
0       Q1   7.05            
1       Q2  22.80          **
2       Q3  45.83        ****
3       Q4  94.14  **********

In [38]:
#create the table
print(f'{'Quartile':<20} {'Adolescent fertility rate (births per 1,000 women ages 15-19)':<65}')
print('_'*18+ ' ' + '_'*80)
print(f'{'':<20} {'Adolescent fertility rate (births per 1,000 women ages 15-19)':<65} {'Sparkline':<65}')
for index, row in quartile_data.iterrows():
    print(f'{row["Quartile"]:<20} {row["2022"]:<65} {row["sparkline"]:<30}')

Quartile             Adolescent fertility rate (births per 1,000 women ages 15-19)    
__________________ ________________________________________________________________________________
                     Adolescent fertility rate (births per 1,000 women ages 15-19)     Sparkline                                                        
Q1                   7.05                                                                                            
Q2                   22.8                                                              **                            
Q3                   45.83                                                             ****                          
Q4                   94.14                                                             **********                    
